File: notes-n-resources/Data-Structures-N-Algo/_DS-n-Algos/__MY_OPRIGINAL_DS/_Extra-Practice/06_breadth-first_search/julia/01_breadth-first_search.jl
using DataStructures, Test, Suppressor

graph = Dict()
graph["you"] = ["alice","bob","claire"]
graph["bob"] = ["anuj","peggy"]
graph["claire"] = ["thom","jonny"]
graph["alice"] = ["peggy"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []

function person_is_seller(name)
	return name[end] == 'm'
end

function search(name)
	search_queue = Deque{String}()
	map(n -> push!(search_queue,n),graph[name])
	searched = []
	while isempty(search_queue) == false
		person = popfirst!(search_queue)
		if (person in searched) == false
			if person_is_seller(person)
				println(person , " is a mango seller")
				return true
			else
				map(n -> push!(search_queue,n),graph[person])
				push!(searched,person)
			end
		end
	end
	return false
end


result = @capture_out search("you")
@test result == "thom is a mango seller
"

File: notes-n-resources/Data-Structures-N-Algo/_DS-n-Algos/__MY_OPRIGINAL_DS/_Extra-Practice/06_breadth-first_search/Golang/01_breadth-first_search.go
package main

func person_is_seller(name string) bool {
	return name[len(name)-1] == 'm'
}

var graph = make(map[string][]string)

func main() {
	graph["you"] = []string{"alice", "bob", "claire"}
	graph["bob"] = []string{"anuj", "peggy"}
	graph["alice"] = []string{"peggy"}
	graph["claire"] = []string{"thom", "jonny"}
	graph["anuj"] = []string{}
	graph["peggy"] = []string{}
	graph["thom"] = []string{}
	graph["jonny"] = []string{}

	search("you")
}

func search(name string) bool {
	var search_queue []string
	search_queue = append(search_queue, graph[name]...)
	var searched []string
	var person string
	for len(search_queue) != 0 {
		person = search_queue[0]
		search_queue = search_queue[1:]
		if person_not_in_searched(person, searched) {
			if person_is_seller(person) {
				println(person + " is mango seller!")
				return true
			}

			search_queue = append(search_queue, graph[person]...)
			searched = append(searched, person)

		}
	}
	return false
}

func person_not_in_searched(person string, searched []string) bool {
	for _, n := range searched {
		if n == person {
			return false
		}
	}
	return true
}
